variables/dependencies.go

```diff
@@ -0,0 +1,129 @@
+type Dependency struct {
+	Name                  string
+	TemplateFolder        string
+	OutputFolder          string
+	DontInheritVariables  bool
+	Variables             []Variable
+}
+
+// Get all the variables in this dependency, namespacing each variable with the name of this dependency
+func (dependency Dependency) GetNamespacedVariables() []Variable {
+	variables := []Variable{}
+
+	for _, variable := range dependency.Variables {
+		variableNameForDependency := fmt.Sprintf("%s.%s", dependency.Name, variable.Name())
+		variables = append(variables, variable.WithName(variableNameForDependency))
+	}
+
+	return variables
+}
+
+// Given a unique variable name, return a tuple that contains the dependency name (if any) and the variable name.
+// Variable and dependency names are split by a dot, so for "foo.bar", this will return ("foo", "bar"). For just "foo",
+// it will return ("", "foo").
+func SplitIntoDependencyNameAndVariableName(uniqueVariableName string) (string, string) {
+	parts := strings.SplitAfterN(uniqueVariableName, ".", 2)
+	if len(parts) == 2 {
+		// The split method leaves the character you split on at the end of the string, so we have to trim it
+		return strings.TrimSuffix(parts[0], "."), parts[1]
+	} else {
+		return "", parts[0]
+	}
+}
+
+// Given a map of key:value pairs read from a Boilerplate YAML config file of the format:
+//
+// dependencies:
+//   - name: <NAME>
+//     template-folder: <TEMPLATE_FOLDER>
+//     output-folder: <OUTPUT_FOLDER>
+//
+//   - name: <NAME>
+//     template-folder: <TEMPLATE_FOLDER>
+//     output-folder: <OUTPUT_FOLDER>
+//
+// This method takes the data above and unmarshals it into a list of Dependency objects
+func UnmarshalDependenciesFromBoilerplateConfigYaml(fields map[string]interface{}) ([]Dependency, error) {
+	unmarshalledDependencies := []Dependency{}
+	dependencyNames := []string{}
+
+	listOfFields, err := unmarshalListOfFields(fields, "dependencies")
+	if err != nil {
+		return unmarshalledDependencies, err
+	}
+
+	for _, fields := range listOfFields {
+		dependency, err := UnmarshalDependencyFromBoilerplateConfigYaml(fields)
+		if err != nil {
+			return unmarshalledDependencies, err
+		}
+
+		if util.ListContains(dependency.Name, dependencyNames) {
+			return unmarshalledDependencies, errors.WithStackTrace(DuplicateDependencyName(dependency.Name))
+		}
+		dependencyNames = append(dependencyNames, dependency.Name)
+
+		unmarshalledDependencies = append(unmarshalledDependencies, *dependency)
+	}
+
+	return unmarshalledDependencies, nil
+}
+
+// Given a map of key:value pairs read from a Boilerplate YAML config file of the format:
+//
+// name: <NAME>
+// template-folder: <TEMPLATE_FOLDER>
+// output-folder: <OUTPUT_FOLDER>
+//
+// This method takes the data above and unmarshals it into a Dependency object
+func UnmarshalDependencyFromBoilerplateConfigYaml(fields map[string]interface{}) (*Dependency, error) {
```

Comment:
The vast majority of time spent in this PR was in this manual unmarshalling code. The lack of a good type system in Go made this more painful than it needed to be.